docs/protocol/core/v3.0.rst

```diff
@@ -0,0 +1,759 @@
+example, the little-endian 32-bit signed integer data type defines
+binary representations for all integers in the range −2,147,483,648 to
+2,147,483,647. The core protocol only considers a limited set of data
+types, but protocol extensions may define other data types.
+
+An array is divided into a set of *chunks*, where each chunk is a
+hyperrectangle defined by a tuple of intervals, one for each dimension
+of the array. The shape of a chunk is the tuple of interval lengths,
+and the size of a chunk (i.e., number of elements contained within the
+chunk) is the product of its interval lengths.
+
+The chunks of an array are organised into a *grid*. The core protocol
+only considers the case where all chunks have the same shape and the
+chunks form a regular grid. However, protocol extensions may define
+other grid types such as rectilinear grids.
+
+An array is associated with a *memory layout* which defines how to
+construct a binary representation of a single chunk by organising the
+binary values within the chunk into a single contiguous sequence of
+bytes. The core protocol defines two types of memory layout based on
+"C" (row-major) and "F" (column-major) ordering of values, but
+protocol extensions may define other memory layouts.
+
+An array is associated with an *encoding pipeline*, which is a
+sequence of zero or more *codecs* that transforms the binary
+representation of a chunk in some way. For example, an encoding
+pipeline might include a checksum codec to ensure data integrity, and
+a compression codec to reduce data size. All codecs implement a common
+*codec interface* which provides a pair of operations, one to perform
+the transformation (encode), the other to reverse the transformation
+(decode).
+
+Each node in a hierarchy is represented by a *metadata document*,
+which is a machine-readable document containing essential processing
+information about the node. For example, an array metadata document
+will specify the number of dimensions, length of each dimension, data
+type, chunk shape, memory layout and encoding pipeline for that array.
+
+The metadata and encoded chunk data for all nodes in a hierarchy are
+held in a *store*. To enable a variety of different store types to be
+used, the core protocol defines a simple *store interface* which is a
+common set of operations that a store must provide.
+
+
+Node names
+----------
+
+Except for the root node, each node in a hierarchy must have a name,
+which is a string of ASCII characters. To ensure consistent behaviour
+across different storage systems, the following constraints apply to
+node names:
+
+* must not be the empty string ("")
+
+* must consist only of characters in the sets `a-z`, `A-Z`, `0-9`,
+  `-_.`
+
+* must not be a string composed only of period characters, e.g. "." or
+  ".."
+
+* must be at most 255 characters long
+
+Note that node names are used to form storage keys, and that some
+storage systems will perform a case-insensitive comparison of storage
```

Comment:
doesn't that mean the storage system is required to perform a case-insensitive comparison of storage keys, in order to have consistent behavior between systems?

Reply:
This is awkward, it's in there because of mac file systems (HFS+, APFS) being case insensitive by default. With the current zarr python implementation we've seen specific issues of bugs being due to two sibling arrays getting created with names like "foo" and "FOO", with data from one ending up in the other. I was trying to come up with something to address this, but it's possibly not the right approach currently.

Note that IIUC mac file systems do case insensitive matching, but do preserve the original case of the file as created. So, e.g., opening "foo.txt" or "FOO.TXT" will open the same file, but if you created it originally as "FOO.TXT" then that's what you'll see when listing a directory.